File: push/src/main/java/com/xiaomi/push/service/JobSchedulerTaskQueue.java
package com.xiaomi.push.service;

final class JobSchedulerTaskQueue {
    private final int defaultHeapSize = 256;
    private int deletedCancelledNumber;
    private int size;
    private JobSchedulerTaskWrapper[] timers = new JobSchedulerTaskWrapper[this.defaultHeapSize];

    private void downHeap(int i) {
        int i2 = i;
        int i3 = (i2 * 2) + 1;
        while (true) {
            int i4 = this.size;
            if (i3 >= i4 || i4 <= 0) {
                return;
            }
            int i5 = i3;
            if (i3 + 1 < i4 && this.timers[i3 + 1].when < this.timers[i3].when) {
                i5 = i3 + 1;
            }
            if (this.timers[i2].when < this.timers[i5].when) {
                return;
            }
            JobSchedulerTaskWrapper[] jobSchedulerTaskWrapperArr = this.timers;
            JobSchedulerTaskWrapper jobSchedulerTaskWrapper = jobSchedulerTaskWrapperArr[i2];
            jobSchedulerTaskWrapperArr[i2] = jobSchedulerTaskWrapperArr[i5];
            jobSchedulerTaskWrapperArr[i5] = jobSchedulerTaskWrapper;
            i3 = (i5 * 2) + 1;
            i2 = i5;
        }
    }

    int getTask(JobSchedulerTaskWrapper jobSchedulerTaskWrapper) {
        for (int i = 0; i < this.timers.length; i++) {
            if (this.timers[i] == jobSchedulerTaskWrapper) {
                return i;
            }
        }
        return -1;
    }

    private void upHeap() {
        int i = this.size - 1;
        int i2 = (i - 1) / 2;
        while (this.timers[i].when < this.timers[i2].when) {
            JobSchedulerTaskWrapper[] jobSchedulerTaskWrapperArr = this.timers;
            JobSchedulerTaskWrapper jobSchedulerTaskWrapper = jobSchedulerTaskWrapperArr[i];
            jobSchedulerTaskWrapperArr[i] = jobSchedulerTaskWrapperArr[i2];
            jobSchedulerTaskWrapperArr[i2] = jobSchedulerTaskWrapper;
            i = i2;
            i2 = (i2 - 1) / 2;
        }
    }

    void adjustMinimum() {
        downHeap(0);
    }

    void delete(int i) {
        if (i >= 0 && i < this.size) {
            JobSchedulerTaskWrapper[] jobSchedulerTaskWrapperArr = this.timers;
            int i2 = this.size - 1;
            this.size = i2;
            jobSchedulerTaskWrapperArr[i] = jobSchedulerTaskWrapperArr[i2];
            jobSchedulerTaskWrapperArr[i2] = null;
            downHeap(i);
        }
    }

    void deleteIfCancelled() {
        int i = 0;
        while (i < this.size) {
            int i2 = i;
            if (this.timers[i].cancelled) {
                this.deletedCancelledNumber++;
                delete(i);
                i2 = i - 1;
            }
            i = i2 + 1;
        }
    }

    boolean hasJob(int i) {
        for (int i2 = 0; i2 < this.size; i2++) {
            if (this.timers[i2].type == i) {
                return true;
            }
        }
        return false;
    }

    boolean hasJob(int i, JobScheduler.Job job) {
        for (int i2 = 0; i2 < this.size; i2++) {
            if (this.timers[i2].job == job) {
                return true;
            }
        }
        return false;
    }

    void insert(JobSchedulerTaskWrapper jobSchedulerTaskWrapper) {
        JobSchedulerTaskWrapper[] jobSchedulerTaskWrapperArr = this.timers;
        int length = jobSchedulerTaskWrapperArr.length;
        int i = this.size;
        if (length == i) {
            JobSchedulerTaskWrapper[] jobSchedulerTaskWrapperArr2 = new JobSchedulerTaskWrapper[i * 2];
            System.arraycopy(jobSchedulerTaskWrapperArr, 0, jobSchedulerTaskWrapperArr2, 0, i);
            this.timers = jobSchedulerTaskWrapperArr2;
        }
        JobSchedulerTaskWrapper[] jobSchedulerTaskWrapperArr3 = this.timers;
        int i2 = this.size;
        this.size = i2 + 1;
        jobSchedulerTaskWrapperArr3[i2] = jobSchedulerTaskWrapper;
        upHeap();
    }

    boolean isEmpty() {
        return this.size == 0;
    }

    JobSchedulerTaskWrapper minimum() {
        return this.timers[0];
    }

    int purge() {
        if (isEmpty()) {
            return 0;
        }
        this.deletedCancelledNumber = 0;
        deleteIfCancelled();
        return this.deletedCancelledNumber;
    }

    void removeJobs(int i) {
        for (int i2 = 0; i2 < this.size; i2++) {
            if (this.timers[i2].type == i) {
                this.timers[i2].cancel();
            }
        }
        deleteIfCancelled();
    }

    void removeJobs(int i, JobScheduler.Job job) {
        for (int i2 = 0; i2 < this.size; i2++) {
            if (this.timers[i2].job == job) {
                this.timers[i2].cancel();
            }
        }
        deleteIfCancelled();
    }

    void reset() {
        this.timers = new JobSchedulerTaskWrapper[this.defaultHeapSize];
        this.size = 0;
    }
}

File: legacy/src/main/java/com/xiaomi/push/service/JobSchedulerTaskQueue.kt
package com.xiaomi.push.service
import io.github.magisk317.mipush.protocol.model.*

class JobSchedulerTaskQueue {
    private val defaultHeapSize = 256
    private var deletedCancelledNumber = 0
    private var size = 0
    private var timers = arrayOfNulls<JobSchedulerTaskWrapper>(defaultHeapSize)

    private fun timerAt(index: Int): JobSchedulerTaskWrapper = checkNotNull(timers[index])

    private fun downHeap(index: Int) {
        var currentIndex = index
        var childIndex = currentIndex * 2 + 1
        while (childIndex < size && size > 0) {
            var candidateIndex = childIndex
            if (childIndex + 1 < size && timerAt(childIndex + 1).`when` < timerAt(childIndex).`when`) {
                candidateIndex = childIndex + 1
            }
            if (timerAt(currentIndex).`when` < timerAt(candidateIndex).`when`) {
                return
            }
            val current = timerAt(currentIndex)
            timers[currentIndex] = timers[candidateIndex]
            timers[candidateIndex] = current
            currentIndex = candidateIndex
            childIndex = currentIndex * 2 + 1
        }
    }

    fun getTask(taskWrapper: JobSchedulerTaskWrapper): Int {
        for (index in timers.indices) {
            if (timers[index] === taskWrapper) {
                return index
            }
        }
        return -1
    }

    private fun upHeap() {
        var index = size - 1
        var parent = (index - 1) / 2
        while (timers[index] != null && timers[parent] != null && timerAt(index).`when` < timerAt(parent).`when`) {
            val current = timerAt(index)
            timers[index] = timers[parent]
            timers[parent] = current
            index = parent
            parent = (index - 1) / 2
        }
    }

    fun adjustMinimum() {
        downHeap(0)
    }

    fun delete(index: Int) {
        if (index in 0 until size) {
            val lastIndex = size - 1
            size = lastIndex
            timers[index] = timers[lastIndex]
            timers[lastIndex] = null
            downHeap(index)
        }
    }

    fun deleteIfCancelled() {
        var index = 0
        while (index < size) {
            var nextIndex = index
            if (timerAt(index).cancelled) {
                deletedCancelledNumber++
                delete(index)
                nextIndex = index - 1
            }
            index = nextIndex + 1
        }
    }

    fun hasJob(type: Int): Boolean {
        for (index in 0 until size) {
            if (timerAt(index).type == type) {
                return true
            }
        }
        return false
    }

    fun hasJob(type: Int, job: JobScheduler.Job): Boolean {
        for (index in 0 until size) {
            if (timerAt(index).job == job) {
                return true
            }
        }
        return false
    }

    fun insert(taskWrapper: JobSchedulerTaskWrapper) {
        if (timers.size == size) {
            val expanded = arrayOfNulls<JobSchedulerTaskWrapper>(size * 2)
            System.arraycopy(timers, 0, expanded, 0, size)
            timers = expanded
        }
        timers[size] = taskWrapper
        size += 1
        upHeap()
    }

    fun isEmpty(): Boolean = size == 0

    fun minimum(): JobSchedulerTaskWrapper = timerAt(0)

    fun purge(): Int {
        if (isEmpty()) {
            return 0
        }
        deletedCancelledNumber = 0
        deleteIfCancelled()
        return deletedCancelledNumber
    }

    fun removeJobs(type: Int) {
        for (index in 0 until size) {
            if (timerAt(index).type == type) {
                timerAt(index).cancel()
            }
        }
        deleteIfCancelled()
    }

    fun removeJobs(type: Int, job: JobScheduler.Job) {
        for (index in 0 until size) {
            if (timerAt(index).job == job) {
                timerAt(index).cancel()
            }
        }
        deleteIfCancelled()
    }

    fun reset() {
        timers = arrayOfNulls(defaultHeapSize)
        size = 0
    }
}
